### PR TITLE
cdp: add TLS capability

### PR DESCRIPTION
**Pull request by lbalaceanu**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [x] PR should be backported to stable branches
- [ ] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

Adding TLS capability to connections between Kamailio and configured peers. Configuration done via kamailio.cfg.

### CODE REVIEW

File: src/modules/cdp/cdp_tls.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+cfg_option_t methods[] = {
+	{"TLSv1",   .val = TLS_USE_TLSv1},
+	{"TLSv1.0", .val = TLS_USE_TLSv1},
+	{"TLSv1+",  .val = TLS_USE_TLSv1_PLUS},
+	{"TLSv1.0+", .val = TLS_USE_TLSv1_PLUS},
+	{"TLSv1.1",  .val = TLS_USE_TLSv1_1},
+	{"TLSv1.1+", .val = TLS_USE_TLSv1_1_PLUS},
+	{"TLSv1.2",  .val = TLS_USE_TLSv1_2},
+	{"TLSv1.2+", .val = TLS_USE_TLSv1_2_PLUS},
```
Comment (oej) [inline src/modules/cdp/cdp_tls.c:11]: TLS 1.3 should be possible

File: src/modules/cdp/cdp_tls.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	/* Create an SSL object */
+	SSL *ssl = SSL_new(ctx);
+	if (!ssl) {
+		LM_ERR("init_ssl_conn(): Failed to create SSL object.");
+		SSL_CTX_free(ctx);
+		return NULL;
+	}
+	/* Set up the TLS connection */
+	if (SSL_set_fd(ssl, client_fd) != 1) {
+		LM_ERR("init_ssl_conn(): Failed to set up TLS connection.");
```
Comment (oej) [inline src/modules/cdp/cdp_tls.c:157]: Can you divulge any more about why it failed? Does the error code hint a reason?

File: src/modules/cdp/cdp_tls.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	}
+	if (certificate.s && private_key.s) {
+		if (load_certificates(*tls_ctx_p, &certificate, &private_key) < 0) {
+			LM_ERR("to_ssl(): Error on loading certificates\n");
+			return -1;
+		}
+	}
+	*tls_conn_p = init_ssl_conn(tcp_sock, *tls_ctx_p);
+	if (!(*tls_conn_p)) {
+		LM_ERR("to_ssl(): Error on negociating ssl connection\n");
```
Comment (oej) [inline src/modules/cdp/cdp_tls.c:192]: TLS connection

File: src/modules/cdp/cdp_tls.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	SSL_shutdown(tls_conn);
+	SSL_free(tls_conn);
+	SSL_CTX_free(tls_ctx);
+}
+
+int to_ssl(SSL_CTX** tls_ctx_p, SSL** tls_conn_p, int tcp_sock, int method)
+{
+	*tls_ctx_p = init_ssl_ctx(method);
+	if (!(*tls_ctx_p)) {
+		LM_ERR("to_ssl(): Error on initialising ssl context\n");
```
Comment (oej) [inline src/modules/cdp/cdp_tls.c:181]: TLS context or "openssl context"

File: src/modules/cdp/doc/cdp_admin.xml [hunk 5]
```diff
@@ -176,10 +176,107 @@ modparam("cdp", "workerq_length_threshold_percentage", 25)
...
+        <programlisting format="linespecific">...
+modparam("tls", "tls_method", "TLSv1")
+...
+	</programlisting>
+      </example>
+ </section>
+ <section>
+      <title>private_key (string)</title>
+
+      <para>Sets the private key file name.</para>
```
Comment (oej) [inline src/modules/cdp/doc/cdp_admin.xml:255]: Do you support both der and pem encoding? Please specify format.

File: src/modules/cdp/doc/cdp_admin.xml [hunk 6]
```diff
@@ -176,10 +176,107 @@ modparam("cdp", "workerq_length_threshold_percentage", 25)
...
+        <programlisting format="linespecific">...
+modparam("cdp", "private_key", "/usr/local/etc/kamailio/my_pkey.pem")
+...
+	</programlisting>
+      </example>
+ </section>
+ <section>
+      <title>certificate (string)</title>
+
+      <para>Sets the certificate file name.</para>
```
Comment (oej) [inline src/modules/cdp/doc/cdp_admin.xml:269]: What type of cert? Signed by whom?

**Comment by oej** [comment]
Just for brainstorming: My general feeling is that a lot of this code - the parsing of the TLS variant, the default settings - should be general for all kamailio modules using TLS. If we decide to deprecate a version of TLS or add a new version, it would be simpler to have it in one place, not spread out over the code base.

Core Kamailio now has two TLS stacks supported. I have no clue, but would it be possible to use those as an interface to get the same multi-stack support here.

**Comment by oej** [comment]
Thank you for your contribution!

**Comment by henningw** [comment]
Thanks also from my side. One additional note, please execute "clang-format" on the file, as the syntax check is currently failing. Just do a force push after you changed it to update the PR commit.

**Comment by lbalaceanu** [comment]
Thank you for the comments, I will make changes accordingly. Will look into the multi-stack issue as well.

**Comment by miconda** [comment]
I think the TLS implementation for SIP can't be used in this module, because the cdp module has its own internal design with dedicated transport workers. Reusing some common functions could be eventually done, but I am not sure how much code is shared in the current approach. The tls config parser is anyhow the ini-parser from core, it is not inside the tls module. At this moment I don't see a straightforward way that having/adding support for a TLS version (say TLS 1.3) in one module becomes available in the other one.

For the moment, the current implementation might have to be merged as it is, future reworking might do a better code sharing.

As a generic note, the name of the function does not have to be explicitly added to the log messages, it is done automatically for all log messages -- it might be that CDP module was not updated to remove function names from old log messages, but the new ones should avoid having function names, they will appear twice.

**Comment by oej** [comment]
Thanks @miconda for the response. I understand.

**Comment by lbalaceanu** [comment]
More changes to follow in subsequent commits. In the end I will squash commits in the PR.

**Comment by lbalaceanu** [comment]
Hi. Hope I took care of the punctual requests made by @oej and @henningw. Please comment. However, starting tomorrow I will not have access to my laptop for a week. Have a nice day!

**Comment by lbalaceanu** [comment]
Hello @oej @henningw @miconda do you think this commit can be merged?

**Comment by carstenbock** [comment]
I did a review - looks good to me, I will merge the PR.

**Comment by lbalaceanu** [comment]
Thank you all for your help.

**Comment by linuxmaniac** [comment]
@lbalaceanu this code is breaking nightly-builds 
I commented at https://github.com/kamailio/kamailio/commit/defae2551f8839c5397523f135293fb2a4056704 with the compile errors
https://kamailio.sipwise.com/job/kamailiodev-nightly-binaries/2489/console

**Pull request merged**